### PR TITLE
fix(anthropic): Only finish relevant spans in .create() patches

### PR DESCRIPTION
**Pull request by alexander-alderman-webb**
### Description

Anthropic spans are only created in `_sentry_patched_create()` and `_sentry_patched_create_async()`.

Exit spans in these functions when the `anthropic` library function raises an exception.

Resolves an edge case where `_sentry_patched_create()` or `_sentry_patched_create_async()` exits early and does not create a span and `get_current_span()` therefore returns a non-anthropic span.

#### Reminders
- Please add tests to validate your changes, and lint your code using `tox -e linters`.
- Add GH Issue ID _&_ Linear ID (if applicable)
- PR title should use [conventional commit](https://develop.sentry.dev/engineering-practices/commit-messages/#type) style (`feat:`, `fix:`, `ref:`, `meta:`)
- For external contributors: [CONTRIBUTING.md](https://github.com/getsentry/sentry-python/blob/master/CONTRIBUTING.md), [Sentry SDK development docs](https://develop.sentry.dev/sdk/), [Discord community](https://discord.gg/Ww9hbqr)

### CODE REVIEW

File: sentry_sdk/integrations/anthropic.py [hunk 1]
```diff
@@ -633,6 +633,7 @@ def _sentry_patched_create_sync(f: "Any", *args: "Any", **kwargs: "Any") -> "Any
         exc_info = sys.exc_info()
         with capture_internal_exceptions():
             _capture_exception(exc)
+            span.__exit__(None, None, None)
```
[the remark this answers is not in the document]
Comment (alexander-alderman-webb) [reply sentry_sdk/integrations/anthropic.py:636]: we have bigger problems if capture_exception raises

[the remark this answers is not in the document]
Comment (alexander-alderman-webb) [reply sentry_sdk/integrations/anthropic.py:636]: duplicate of https://github.com/getsentry/sentry-python/pull/5716#discussion_r3026787436

**Comment by alexander-alderman-webb** [comment]
> Is there any test coverage that we can add here to confirm this behaviour? Otherwise LGTM

In my view a regression test has limited use in this case because an integration should not be touching spans created outside the integration (unless it's for an agreed-upon reason, i.e., not duplicating spans between `openai-agents` and `openai`).

For the future, here's a script that verifies that demonstrates the edge case I'm addressing below. The `messages` argument is missing from the call to `create()` and the outer transaction is closed prematurely since it's status is `INTERNAL_ERROR`.

```python
import sentry_sdk
from sentry_sdk.tracing import SPANSTATUS
from sentry_sdk.integrations.anthropic import AnthropicIntegration
from anthropic import Anthropic

sentry_sdk.init(
    dsn="",
    integrations=[AnthropicIntegration()],
    traces_sample_rate=1.0,
)

def main():
    client = Anthropic(api_key="dummy")
    with sentry_sdk.start_transaction(name="parent", op="test") as transaction:
        transaction.set_status(SPANSTATUS.INTERNAL_ERROR)
        try:
            client.messages.create(max_tokens=1024, model="model", stream=False)
        except Exception as e:
            print(f"Expected exception: {e}")

if __name__ == "__main__":
    main()
```

**Comment by ericapisani** [comment]
> In my view a regression test has limited use in this case because an integration should not be touching spans created outside the integration (unless it's for an agreed-upon reason, i.e., not duplicating spans between openai-agents and openai).

I agree with what you're saying regarding an integration not touching spans created outside of it, but I think we could ensure that this edge case is resolved with a test that confirms that an anthropic span is present when the conditions that caused it are set.

To give a picture of what I'm envisioning (I used the script you've provided as the basis for it):

```python
    sentry_init(integrations=[AnthropicIntegration()], traces_sample_rate=1.0)
    events = capture_events()

    client = Anthropic(api_key="z")
    client.messages._post = mock.Mock(return_value=EXAMPLE_MESSAGE)

    with start_transaction(name="anthropic") as transaction:
        transaction.set_status(SPANSTATUS.INTERNAL_ERROR)
        client.messages.create(
           # No messages value passed in
            model="model",
            max_tokens=1024,
            stream=False
        )

    (tx,) = events

    assert len(tx["spans"]) == 1
    span = tx["spans"][0]

    assert span["data"][SPANDATA.GEN_AI_SYSTEM] == "anthropic"
```

What do you think? Also happy to pair on this tomorrow when you're back online 😄
